### ISSUE TITLE
Add unit conversion like google.

### ISSUE THREAD

**Issue by deam0n** [searxng/searxng#3015]
**Is your feature request related to a problem? Please describe.**
no

**Describe the solution you'd like**
Unit conversion like google when performing searched like "10 miles in km"

**Describe alternatives you've considered**
use google.com

**Additional context**
Currency conversion already works, however distance, weight, speed would be nice to have it implemented.

**Comment by thunder-coding** [comment]
Looks like ddg_definitions does handle unit conversion, but SearXNG engine crashes when it tries to parse the result.

Search query: `!ddg_definitions km to miles`
Error produced: 

```
Error
  * Error: httpx.ConnectTimeout
  * Percentage: 0
  * Parameters: `(None, None, 'api.duckduckgo.com')`
  * File name: `searx/search/processors/online.py:118`
  * Function: `_send_http_request`
  * Code: `response = req(params['url'], **request_args)`

Error
  * Error: AttributeError
  * Percentage: 0
  * Parameters: `()`
  * File name: `searx/engines/duckduckgo_definitions.py:105`
  * Function: `response`
  * Code: `results.append({'answer': html_to_text(answer), 'url': search_res.get('AbstractURL', '')})`
  ```

**Comment by return42** [comment]
before we start implementing too much, it might be easier to use wolframalpha

- [!wa 10 miles in km](https://darmarit.org/searx/search?q=%21wa+10+miles+in+km)
- [!!wa 10 miles in km](https://darmarit.org/searx/search?q=%21%21wa+10+miles+in+km)

@thunder-coding can you please open an issue / thanks!

**Comment by Bnyro** [comment]
The only two ways we could really achieve this without having to do all the logic ourselves seem to be using
- using the [pint](https://github.com/hgrecco/pint) library
- [GNU units](https://www.gnu.org/software/units/manual/units.html) via a subprocess

I personally think that it'd be okay to depend on the python units library, using GNU units would be more flexible and supports more currencies but running it in a subprocess is probably not really a good solution?

What are your opinions on that @dalf, @unixfox and @return42?

**Comment by dalf** [comment]
I don't have a clear answer right now.
Here some words, perhaps not in the right direction:

> without having to do all the logic ourselves seem

A naive (?) solution:
* There is already a [JSON file with all the unit from wikidata](https://github.com/searxng/searxng/blob/master/searx/data/wikidata_units.json). The [existing SPARQL query](https://github.com/searxng/searxng/blob/e2af3e49702f6fb40e1614f826544dc3b03bca2f/searxng_extra/update/update_wikidata_units.py#L35-L44). 
* Updated query to get the conversion to SI unit: https://w.wiki/9VFr
* From that we can implement something to convert unit:
    * "source unit" to SI
    * SI to "destination unit"

---

More or less related to this 9 years old PR: https://github.com/searx/searx/pull/384
Especially this: https://docs.sympy.org/latest/modules/physics/units/index.html

---

[EDIT] There is https://github.com/Qalculate/libqalculate but there is no Python binding.
<details>

<summary>untested binding written by ChatGPT4</summary>

based on http://qalculate.github.io/reference/index.html

qalculate_wrapper.pyx:

```python
# distutils: language = c++
# distutils: sources = qalculate/Calculator.cpp

cimport cython

# External C++ class declaration
cdef extern from "qalculate/Calculator.h" namespace "libqalculate":
    cdef cppclass Calculator:
        Calculator() except +
        void loadExchangeRates()
        void loadGlobalDefinitions()
        void loadLocalDefinitions()
        string calculateAndPrint(string str, int msecs, const EvaluationOptions &eo, const PrintOptions &po, string *parsed_expression)

# Cython wrapper for the Calculator class
cdef class PyCalculator:
    cdef Calculator* c_calculator

    def __cinit__(self):
        self.c_calculator = new Calculator()
        self.c_calculator.loadExchangeRates()
        self.c_calculator.loadGlobalDefinitions()
        self.c_calculator.loadLocalDefinitions()

    def __dealloc__(self):
        del self.c_calculator

    def calculate_and_print(self, expression, int msecs=10000):
        cdef string result
        cdef string c_expression = expression.encode('UTF-8')
        result = self.c_calculator.calculateAndPrint(c_expression, msecs, default_user_evaluation_options, default_print_options, NULL)
        return result.decode('UTF-8')
```

setup.py
```python
from setuptools import setup
from Cython.Build import cythonize

setup(
    ext_modules=cythonize("qalculate_wrapper.pyx",
                          language_level="3"),
    compiler_directives={'language_level': "3"}
)
```

</details>



---

[EDIT2] [Spot](https://gitlab.e.foundation/e/infra/spot) uses [mathjs](https://mathjs.org/index.html)

**Comment by thunder-coding** [comment]
> GNU units via a subprocess

Terrible idea, increases attack surface for instances for RCE vulnerabilities.
